### ISSUE TITLE
boxed_local doesn't work with the FnBox pattern.

### ISSUE THREAD

**Issue by Stebalien** [rust-lang-nursery/rust-clippy#916]
In my template library horrorshow, I have the following definition to allow `RenderOnce::render_once` to be called through a boxed `RenderOnce`.

``` rust
pub trait RenderBox {
    /// Do not call. Called by RenderOnce impl on Box<RenderBox>
    #[doc(hidden)]
    fn render_box(self: Box<Self>, tmpl: &mut TemplateBuffer);

    /// Do not call. Called by RenderOnce impl on Box<RenderBox>
    #[doc(hidden)]
    fn size_hint_box(&self) -> usize;
}


impl<T> RenderBox for T where T: RenderOnce {
    fn render_box(self: Box<T>, tmpl: &mut TemplateBuffer) {
        (*self).render_once(tmpl);
    }

    fn size_hint_box(&self) -> usize {
        RenderOnce::size_hint(self)
    }
}

// Box<RenderBox>

impl<'b> RenderOnce for Box<RenderBox + 'b> {
    #[inline]
    fn render_once(self, tmpl: &mut TemplateBuffer) {
        RenderBox::render_box(self, tmpl);
    }

    #[inline]
    fn size_hint(&self) -> usize {
        RenderBox::size_hint_box(self)
    }
}
```

This is the same pattern as stdlib's `FnBox`/`FnOnce`. Unfortunately, clippy claims that this box is unnecessary:

``` rust
src/render.rs:68:19: 68:23 warning: local variable doesn't need to be boxed here, #[warn(boxed_local)] on by default
src/render.rs:68     fn render_box(self: Box<T>, tmpl: &mut TemplateBuffer) {
```

**Comment by oli-obk** [comment]
well... it *is* unnecessary at the implementation site unless you also add a `?Sized` bound.

**Comment by Stebalien** [comment]
`RenderBox` must only be implemented on `Box<T: Sized>` because it moves a value of type `T` out of the box. Because `RenderBox` knows the size of the real type, `RenderOnce` can be implemented on `Box<RenderBox>` to recover the size and move the real `T` out of the box.

See FnBox: https://doc.rust-lang.org/src/alloc/up/src/liballoc/boxed.rs.html#529-533

**Comment by phansch** [comment]
Fixed via #3321 (feel free to re-open if it's still a problem)
